firedrake/ensemble/ensemble_function.py

```diff
@@ -0,0 +1,292 @@
+        else:
+            for us in self.subfunctions:
+                us *= other
+        return self
+
+    @PETSc.Log.EventDecorator()
+    def __add__(self, other):
+        new = self.copy()
+        new += other
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def __mul__(self, other):
+        new = self.copy()
+        new *= other
+        return new
+
+    @PETSc.Log.EventDecorator()
+    def __rmul__(self, other):
+        return self.__mul__(other)
+
+    @contextmanager
+    def vec(self):
+        """
+        Context manager for the global PETSc Vec with read/write access.
+
+        It is invalid to access the Vec outside of a context manager.
+        """
+        # _fbuf.vec shares the same storage as _vec, so we need this
+        # nested context manager so that the data gets copied to/from
+        # the Function.dat storage and _vec.
+        # However, this copy is done without _vec knowing, so we have
+        # to manually increment the state.
+        with self._fbuf.dat.vec:
+            self._vec.stateIncrease()
+            yield self._vec
+
+    @contextmanager
+    def vec_ro(self):
+        """
+        Context manager for the global PETSc Vec with read only access.
+
+        It is invalid to access the Vec outside of a context manager.
+        """
+        # _fbuf.vec shares the same storage as _vec, so we need this
+        # nested context manager to make sure that the data gets copied
+        # to the Function.dat storage and _vec.
+        with self._fbuf.dat.vec_ro:
+            self._vec.stateIncrease()
+            yield self._vec
+
+    @contextmanager
+    def vec_wo(self):
+        """
+        Context manager for the global PETSc Vec with write only access.
+
+        It is invalid to access the Vec outside of a context manager.
+        """
+        # _fbuf.vec shares the same storage as _vec, so we need this
+        # nested context manager to make sure that the data gets copied
+        # from the Function.dat storage and _vec.
+        with self._fbuf.dat.vec_wo:
+            yield self._vec
+
+
+class EnsembleFunction(EnsembleFunctionBase):
+    """
+    A mixed finite element Function distributed over an ensemble.
+
+    Parameters
+    ----------
+
+    function_space : :class:`EnsembleFunctionSpace`
+        The function space of the function.
+    """
+    def __new__(cls, function_space):
+        if isinstance(function_space, EnsembleDualSpace):
+            return EnsembleCofunction(function_space)
+        return super().__new__(cls)
+
+    def __init__(self, function_space):
+        if not isinstance(function_space, EnsembleFunctionSpace):
+            raise TypeError(
+                "EnsembleFunction must be created using an EnsembleFunctionSpace")
+        super().__init__(function_space)
+
+    def norm(self, *args, **kwargs):
+        """
+        Compute the norm of the function.
+        Any arguments are forwarded to :function:`firedrake.norm`.
```

Comment:
I think numpydoc would say to write this as
```suggestion
        """Compute the norm of the function.
        
        Any arguments are forwarded to `firedrake.norm`.
```